### PR TITLE
parser: add support for `ALTER DATABASE` syntax

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
Add support for `ALTER DATABASE` syntax

Related issue: https://github.com/pingcap/tidb/issues/4641

### What is changed and how it works?
Modified parser.y

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to update the documentation

### CODE REVIEW

File: ast/ddl.go [hunk 1]
```diff
@@ -118,6 +118,42 @@ func (n *CreateDatabaseStmt) Accept(v Visitor) (Node, bool) {
...
+	ctx.WriteKeyWord("ALTER DATABASE")
+	if n.Name != "" {
+		ctx.WritePlain(" ")
+		ctx.WriteName(n.Name)
+	}
+	for _, option := range n.Options {
+		ctx.WritePlain(" ")
+		err := option.Restore(ctx)
+		if err != nil {
+			return errors.Trace(err)
```
Comment (kennytm) [inline ast/ddl.go]: We typically use `errors.Annotate()` here

File: parser_test.go [hunk 2]
```diff
@@ -1844,7 +1844,38 @@ func (s *testParserSuite) TestDDL(c *C) {
...
+		{"ALTER DATABASE DEFAULT CHARSET = 'utf8mb4' COLLATE = 'utf8_bin'", true, "ALTER DATABASE CHARACTER SET = utf8mb4 COLLATE = utf8_bin"},
+		{
+			"ALTER DATABASE DEFAULT CHARSET = 'utf8mb4' DEFAULT COLLATE = 'utf8mb4_general_ci' CHARACTER SET = 'utf8' COLLATE = 'utf8mb4_bin'",
+			true,
+			"ALTER DATABASE CHARACTER SET = utf8mb4 COLLATE = utf8mb4_general_ci CHARACTER SET = utf8 COLLATE = utf8mb4_bin",
+		},
+
+		{"ALTER TABLE ADD COLUMN (a SMALLINT UNSIGNED)", false, ""},
+		{"ALTER TABLE t ADD COLUMN (a SMALLINT UNSIGNED, b varchar(255))", true, "ALTER TABLE `t` ADD COLUMN (`a` SMALLINT UNSIGNED, `b` VARCHAR(255))"},
+		{"ALTER TABLE t ADD COLUMN (a SMALLINT UNSIGNED FIRST)", false, ""},
```
Comment (kennytm) [inline parser_test.go]: These three seems irrelevant and duplicating the tests below.

File: ast/ddl.go (hunk 1 above)
Comment (bb7133) [reply ast/ddl.go]: addressed, thanks

File: parser_test.go (hunk 2 above)
Comment (bb7133) [reply parser_test.go]: done

**Comment by bb7133** [comment]
PTAL @kennytm , thank you!

File: parser.y [hunk 3]
```diff
@@ -606,7 +606,8 @@ import (
 
 %type	<statement>
 	AdminStmt			"Check table statement or show ddl statement"
-	AlterTableStmt			"Alter table statement"
+	AlterDatabaseStmt		"Alter table statement"
+	AlterTableStmt			"Alter database statement"
```
Comment (kennytm) [inline parser.y]: These two are flipped.

File: parser.y [hunk 4]
```diff
@@ -2036,6 +2040,36 @@ IndexColNameList:
...
+ *
+ *  ALTER {DATABASE | SCHEMA} [db_name]
+ *   alter_specification ...
+ *
+ *  alter_specification:
+ *   [DEFAULT] CHARACTER SET [=] charset_name
+ * | [DEFAULT] COLLATE [=] collation_name
+ *******************************************************************************************/
+ AlterDatabaseStmt:
+	"ALTER" DatabaseSym DBNameOpt DatabaseOptionListOpt
```
Comment (kennytm) [inline parser.y]: I think it's better to split into two rules rather than passing an `[]interface{}`:

```
    "ALTER" DatabaseSym DBName DatabaseOptionList
|   "ALTER" DatabaseSym DatabaseOptionList
```

BTW why use a `DatabaseOptionListOpt` and then check `len(opts) == 0` instead of using `DatabaseOptionList`?

File: parser.y (hunk 3 above)
Comment (bb7133) [reply parser.y]: Yep... you are so careful

File: parser.y [hunk 5]
```diff
@@ -3237,7 +3270,7 @@ Identifier:
 identifier | UnReservedKeyword | NotKeywordToken | TiDBKeyword
 
 UnReservedKeyword:
- "ACTION" | "ASCII" | "AUTO_INCREMENT" | "AFTER" | "ALWAYS" | "AVG" | "BEGIN" | "BIT" | "BOOL" | "BOOLEAN" | "BTREE" | "BYTE" | "CLEANUP" | "CHARSET"
+ "ACTION" | "ASCII" | "AUTO_INCREMENT" | "AFTER" | "ALWAYS" | "AVG" | "BEGIN" | "BIT" | "BOOL" | "BOOLEAN" | "BTREE" | "BYTE" | "CLEANUP" | "CHARSET" %prec charsetKwd
```
Comment (zimulala) [inline parser.y]: Do we need to care about `"CHARACTER" "SET"`? I am not very clear about this.

Comment (winkyao) [reply parser.y]: why we use `%prec` here?

Comment (bb7133) [reply parser.y]: > Do we need to care about `"CHARACTER" "SET"`? I am not very clear about this.

@zimulala  The reason is `CHARACTER` is a reserved token but `CHARSET` is unreserved. For example, 

```
mysql> create database charset;
Query OK, 1 row affected

mysql> create database character;
(1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'character' at line 1")
```

Comment (bb7133) [reply parser.y]: > why we use `%prec` here?

@zimulala @winkyao the reason is `CHARSET` is an unreserved keyword. There's a shift/reduce conflict when parsing `AlterDatabseStmt CHARSET`, YACC need to decide if charset is treated as `CHARACTER SET` keyword, or the database name.

BTW, this precedence leads to the following parsing result:

```
mysql> alter database charset=utf8;
(1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=utf8' at line 1")

mysql> alter database character set=utf8;
Query OK, 1 row affected
```
And it is consistent with MySQL parser.
